### PR TITLE
Implement test suite

### PR DESCRIPTION
**Pull request by ryndubei**
8f2d91e8ddf89bae9556cc9d1e0a6ab8ed0fe3a8 implements a test suite which can be run with `cabal test`. At the moment, there are 3 tests, of which 2 (`NumInstance` and `DeriveGeneric`) are supposed to fail, in preparation for adding support for type class instances to Weeder.

A test `ExampleTest` can be easily added by creating `ExampleTest.stdout`, `ExampleTest.dhall` and a directory `ExampleTest` in `test/Spec` containing the modules to be compiled to `hie` files, and specifying the modules in `weeder.cabal`.

The rest of the commits allow the test suite to export the dependency graph of each test to a dotfile, and to convert those dotfiles to PNG images with GraphViz when given the `--graphviz` flag, for example as `cabal test --test-option=--graphviz`, for easier debugging.

Graph of BasicExample.hs

### CODE REVIEW

**Comment by ryndubei** [comment]
Test suite converted to `toml` and failing tests removed for now, should be good to be merged.

File: src/Weeder/Main.hs [hunk 1]
```diff
@@ -105,7 +105,15 @@ main = do
...
-mainWithConfig hieExt hieDirectories requireHsFiles Config{ rootPatterns, typeClassRoots } = do
+mainWithConfig hieExt hieDirectories requireHsFiles weederConfig = 
+  mainWithConfig' hieExt hieDirectories requireHsFiles weederConfig 
+    >>= \(success, _) -> unless success exitFailure
+
+
+-- | Alternative mainWithConfig that returns the analysis and whether a zero exit
+-- code should be returned.
+mainWithConfig' :: String -> [FilePath] -> Bool -> Config -> IO (Bool, Analysis)
+mainWithConfig' hieExt hieDirectories requireHsFiles Config{ rootPatterns, typeClassRoots } = do
```
Comment (ocharles) [inline src/Weeder/Main.hs:116]: I'm not very big on `'` naming, and we now end up with three `main*` functions. This seems a bit confusing. How about we go with just having `mainWithConfig` return `(ExitCode, Analysis)`. Then `main` can be responsible for calling `exitWith`, and your test suite can check `ExitCode` too.

So to recap:

* `main` calls `mainWithConfig` and then calls `exitWith` with the `ExitCode` that `mainWithConfig` returns
* `mainWithConfig` reurns `(ExitCode, Analysis)`. The `ExitCode` is formatted based on `null warnings`, as before.
* We drop the new `mainWithConfig'`.

**Review by ocharles — CHANGES_REQUESTED** [review]

**Comment by ocharles** [comment]
LGTM!

**Comment by ryndubei** [comment]
Should I merge this then?

**Comment by ocharles** [comment]
Yep, go ahead

**Pull request merged**